libz-rs-sys/src/gz.rs

```diff
@@ -521,89 +520,127 @@ unsafe fn deallocate_cstr(s: *mut c_char) {
 /// This function may be called at most once for any file handle.
 #[cfg_attr(feature = "export-symbols", export_name = crate::prefix!(gzclose))]
 pub unsafe extern "C-unwind" fn gzclose(file: gzFile) -> c_int {
-    let ret = match unsafe { file.cast::<GzState>().as_mut() } {
-        Some(state) => match state.mode {
-            GzMode::GZ_READ => unsafe { gzclose_r(state) },
-            GzMode::GZ_WRITE | GzMode::GZ_APPEND | GzMode::GZ_NONE => gzclose_w(state),
-        },
-        None => Z_STREAM_ERROR,
+    let mode = match unsafe { file.cast::<GzState>().as_ref() } {
+        // Note: It is important that this `state` reference lives in a nested scope
+        // because its lifetime must not overlap the calls to `gzclose_r`/`gzclose_w` below,
+        // which free the referenced memory.
+        Some(state) => state.mode,
+        None => {
+            return Z_STREAM_ERROR;
+        }
     };
 
-    // Note: In zlib-ng, the gzclose_r and gzclose_w functions deallocate the state
-    // object themselves. Here, since those functions take references to the state, the
-    // deallocation cannot safely happen until the references are out of scope. Therefore
-    // the free_state call handles the deallocation using the raw pointer to the state.
-    unsafe { free_state(file.cast::<GzState>()) };
-
-    ret
+    match mode {
+        GzMode::GZ_READ => unsafe { gzclose_r(file) },
+        _ => unsafe { gzclose_w(file) },
```

Comment:
may as well make this exhaustive right? 

```suggestion
        GzMode::GZ_WRITE | GzMode::GZ_APPEND | GzMode::GZ_NONE => unsafe { gzclose_w(file) },
```